File: web/src/lib/behaviors.ts
// Behaviors: what the software must do, the tests that show it, and the runs
// that report on those tests.
//
// A behavior is written for people and may point at a section of the plan. A
// test is only a key (`playwright:editor.spec.ts › keeps edits`) linked to one
// or more behaviors; Takomo never stores test code. A run is one report — from
// CI or an agent — of outcomes per test key. A behavior's status is computed by
// the server from the latest result of each linked test; nothing sets it.
// Timestamps are ISO-8601 strings, as everywhere else in the API.
import { api } from './api'
import { fmtAge } from './format'

export const BEHAVIOR_STATUSES = ['failing', 'stale', 'untested', 'verified'] as const
export type BehaviorStatus = (typeof BEHAVIOR_STATUSES)[number]
export type Outcome = 'pass' | 'fail'

export interface LastResult {
  test: string
  outcome: Outcome
  at: string
  commit: string | null
  run: string
}

export interface Behavior {
  id: string
  project: string
  /** A plan node id, or null for a behavior about no section in particular. */
  section: string | null
  title: string
  statement: string
  /** Linked test keys, sorted. */
  tests: string[]
  status: BehaviorStatus
  last_result: LastResult | null
  created_by: string
  created_at: string
  updated_at: string
}

export interface TestResult {
  test: string
  latest: {
    outcome: Outcome
    detail: string | null
    at: string
    commit: string | null
    run: string
    actor: string
  } | null
}

export interface HistoryEntry {
  test: string
  outcome: Outcome
  detail: string | null
  at: string
  commit: string | null
  run: string
  note: string | null
  actor: string
}

export interface BehaviorDetail extends Behavior {
  test_results: TestResult[]
  /** Newest first, at most 50. */
  history: HistoryEntry[]
}

export interface StatusCounts {
  total: number
  verified: number
  failing: number
  stale: number
  untested: number
}

export interface Run {
  id: string
  project: string
  commit: string | null
  note: string | null
  actor: string
  at: string
  passed: number
  failed: number
}

export interface UnlinkedTest {
  test: string
  outcome: Outcome
  at: string
  commit: string | null
}

export interface VerificationSummary {
  fresh_days: number
  summary: StatusCounts
  /** Counts per plan node id. */
  sections: Record<string, StatusCounts>
  unsectioned: number
  unlinked_tests: { items: UnlinkedTest[]; total: number; limit: number }
  latest_run: Run | null
}

export interface Paged<T> {
  items: T[]
  total: number
  limit: number
  note?: string
}

export interface BehaviorFields {
  title: string
  statement?: string
  section?: string | null
  tests?: string[]
}

const json = { 'Content-Type': 'application/json' }
const enc = encodeURIComponent

export function listBehaviors(
  token: string,
  project: string,
  filter: { section?: string; status?: BehaviorStatus; q?: string; limit?: number; offset?: number } = {},
): Promise<Paged<Behavior>> {
  const query = new URLSearchParams()
  for (const [key, value] of Object.entries(filter)) {
    if (value !== undefined && value !== '') query.set(key, String(value))
  }
  const tail = query.toString() ? `?${query}` : ''
  return api<Paged<Behavior>>(token, `/projects/${enc(project)}/behaviors${tail}`)
}

export function fetchVerification(token: string, project: string): Promise<VerificationSummary> {
  return api<VerificationSummary>(token, `/projects/${enc(project)}/verification`)
}

export function createBehavior(token: string, project: string, fields: BehaviorFields): Promise<Behavior> {
  return api<Behavior>(token, `/projects/${enc(project)}/behaviors`, {
    method: 'POST',
    headers: json,
    body: JSON.stringify(fields),
  })
}

export function getBehavior(token: string, id: string): Promise<BehaviorDetail> {
  return api<BehaviorDetail>(token, `/behaviors/${enc(id)}`)
}

/** `section: null` clears the link; `tests` replaces the full list. */
/** A behavior edit. `tests` replaces the whole list; prefer `add_tests` and
 * `remove_tests`, which the server applies against the current list, so a
 * link someone else makes at the same moment is kept. */
export type BehaviorPatch = Partial<BehaviorFields> & { add_tests?: string[]; remove_tests?: string[] }

export function patchBehavior(token: string, id: string, fields: BehaviorPatch): Promise<Behavior> {
  return api<Behavior>(token, `/behaviors/${enc(id)}`, {
    method: 'PATCH',
    headers: json,
    body: JSON.stringify(fields),
  })
}

export function deleteBehavior(token: string, id: string): Promise<unknown> {
  return api(token, `/behaviors/${enc(id)}`, { method: 'DELETE' })
}

/** One test key per line; blank lines and surrounding whitespace dropped, duplicates removed. */
export function parseTestKeys(text: string): string[] {
  return [...new Set(text.split('\n').map((line) => line.trim()).filter(Boolean))]
}

/** "a1b2c3d · 3d" — a short commit when known, then the age. */
export function resultStamp(result: { commit: string | null; at: string }, now: number = Date.now()): string {
  const age = fmtAge(result.at, now)
  return result.commit ? `${result.commit.slice(0, 7)} · ${age}` : age
}

File: src/api/behaviors.rs
//! Verification — `/v1/projects/{project}/behaviors`, `/v1/behaviors/{id}`,
//! `/v1/projects/{project}/runs` and `/v1/projects/{project}/verification`.
//!
//! Takomo stores; the reporter computes. CI or an agent runs the tests and
//! reports pass/fail per test key; this layer validates shapes and scope, and
//! the store derives each behavior's status from what was reported. See
//! `docs/verification.md`.

use super::{
    body_object, first, get_str, get_string_array, parse_i64_param, query_pairs, reject_unknown,
    require_str, ApiJson,
};
use crate::auth::AuthCtx;
use crate::error::{ApiError, ApiResult};
use crate::server::AppState;
use crate::store::{
    BehaviorCreate, BehaviorFilter, BehaviorPatch, ResultInput, RunReport, MAX_BEHAVIORS_PAGE,
    MAX_RUNS_PAGE,
};
use axum::extract::{Path, RawQuery, State};
use axum::http::{HeaderMap, StatusCode};
use axum::response::IntoResponse;
use axum::{Extension, Json};
use serde_json::{Map, Value};
use std::sync::Arc;

const CREATE_FIELDS: [&str; 4] = ["title", "statement", "section", "tests"];
const PATCH_FIELDS: [&str; 4] = ["title", "statement", "section", "tests"];
const RUN_FIELDS: [&str; 3] = ["commit", "note", "results"];
const RESULT_FIELDS: [&str; 3] = ["test", "outcome", "detail"];
const MAX_IDEMPOTENCY_HEADER: usize = 128;

/// The section must be a node of this project's plan.
///
/// Checked here rather than in the store because sections live in the plan's
/// CRDT document, not in a table, and only this layer can open it. A project
/// holds at most one plan, which is what makes a bare node id resolvable.
pub async fn validate_section(
    state: &Arc<AppState>,
    project: &str,
    section: &str,
) -> ApiResult<()> {
    let index = crate::api::mindmaps::project_node_index(state, project).await?;
    if index.iter().any(|(_, n)| n.id == section) {
        return Ok(());
    }
    Err(ApiError::validation(
        "validation.behavior_section",
        format!(
            "No section '{section}' in this project's plan. A behavior names the part of the \
             specification it comes from, so the id has to be one of its sections."
        ),
    )
    .remedy(
        "Read the plan with takomo_plan_read or GET /v1/mindmaps/{id} — every section carries \
         its node id — or leave `section` out for a behavior the specification does not state yet."
            .to_string(),
    ))
}

/// A field that may be absent (leave alone), null (clear) or a string (set).
fn nullable_str(obj: &Map<String, Value>, key: &str) -> ApiResult<Option<Option<String>>> {
    match obj.get(key) {
        None => Ok(None),
        Some(Value::Null) => Ok(Some(None)),
        Some(Value::String(s)) if s.trim().is_empty() => Ok(Some(None)),
        Some(Value::String(s)) => Ok(Some(Some(s.trim().to_string()))),
        Some(_) => Err(ApiError::validation(
            "validation.behavior_section",
            format!("'{key}' must be a plan node id string or null."),
        )
        .remedy("Send a section id, or null to unlink it.".to_string())),
    }
}

fn idempotency_key(headers: &HeaderMap) -> ApiResult<Option<String>> {
    let key = headers
        .get("Idempotency-Key")
        .and_then(|v| v.to_str().ok())
        .map(str::trim)
        .filter(|k| !k.is_empty())
        .map(str::to_string);
    if key
        .as_ref()
        .is_some_and(|k| k.len() > MAX_IDEMPOTENCY_HEADER)
    {
        return Err(ApiError::bad_request(
            "validation.idempotency_key",
            "Idempotency-Key must be at most 128 characters.",
        ));
    }
    Ok(key)
}

/// POST /v1/projects/{project}/behaviors (write).
pub async fn create(
    State(state): State<Arc<AppState>>,
    Extension(ctx): Extension<AuthCtx>,
    Path(project): Path<String>,
    ApiJson(body): ApiJson<Value>,
) -> ApiResult<impl IntoResponse> {
    ctx.require_scope("write")?;
    ctx.require_project(&project)?;
    let obj = body_object(&body)?;
    reject_unknown(obj, &CREATE_FIELDS)?;
    let section = nullable_str(obj, "section")?.flatten();
    if let Some(s) = &section {
        validate_section(&state, &project, s).await?;
    }
    let req = BehaviorCreate {
        project,
        title: require_str(obj, "title")?,
        statement: get_str(obj, "statement")?.unwrap_or_default(),
        section,
        tests: get_string_array(obj, "tests")?.unwrap_or_default(),
    };
    let behavior = state.store.create_behavior(&req, &ctx.actor)?;
    state.wake();
    Ok((StatusCode::CREATED, Json(behavior.to_json())))
}

/// GET /v1/projects/{project}/behaviors?section=&status=&q=&limit=&offset= (read).
///
/// `section=none` narrows to behaviors tied to no section — the ones the
/// specification does not state yet.
pub async fn list(
    State(state): State<Arc<AppState>>,
    Extension(ctx): Extension<AuthCtx>,
    Path(project): Path<String>,
    RawQuery(raw): RawQuery,
) -> ApiResult<Json<Value>> {
    ctx.require_scope("read")?;
    ctx.require_project(&project)?;
    let pairs = query_pairs(raw.as_deref());
    let filter = BehaviorFilter {
        project,
        section: first(&pairs, "section").map(|s| {
            if s == "none" {
                String::new()
            } else {
                s.to_string()
            }
        }),
        status: first(&pairs, "status").map(str::to_string),
        q: first(&pairs, "q").map(str::to_string),
        limit: parse_i64_param(&pairs, "limit")?,
        offset: parse_i64_param(&pairs, "offset")?,
    };
    let limit = filter
        .limit
        .unwrap_or(MAX_BEHAVIORS_PAGE)
        .clamp(1, MAX_BEHAVIORS_PAGE);
    let store = state.clone();
    let (items, total) = super::blocking_read(move || store.store.list_behaviors(&filter)).await?;
    Ok(Json(super::paged(
        items.iter().map(|b| b.to_json()).collect(),
        total,
        limit,
        "Page with ?offset=N, or narrow with ?section=, ?status= or ?q=.",
    )))
}

/// GET /v1/behaviors/{id} (read) — the behavior, each linked test's latest
/// result, and recent history.
pub async fn get(
    State(state): State<Arc<AppState>>,
    Extension(ctx): Extension<AuthCtx>,
    Path(id): Path<String>,
) -> ApiResult<Json<Value>> {
    ctx.require_scope("read")?;
    let existing = state.store.get_behavior(&id)?;
    ctx.require_project(&existing.project)?;
    Ok(Json(state.store.behavior_detail(&id)?))
}

/// PATCH /v1/behaviors/{id} (write). `section: null` unlinks it; `tests`
/// replaces the whole list.
pub async fn patch(
    State(state): State<Arc<AppState>>,
    Extension(ctx): Extension<AuthCtx>,
    Path(id): Path<String>,
    ApiJson(body): ApiJson<Value>,
) -> ApiResult<Json<Value>> {
    ctx.require_scope("write")?;
    let existing = state.store.get_behavior(&id)?;
    ctx.require_project(&existing.project)?;
    let obj = body_object(&body)?;
    reject_unknown(obj, &PATCH_FIELDS)?;
    let section = nullable_str(obj, "section")?;
    if let Some(Some(s)) = &section {
        validate_section(&state, &existing.project, s).await?;
    }
    let patch = BehaviorPatch {
        title: get_str(obj, "title")?,
        statement: get_str(obj, "statement")?,
        section,
        tests: get_string_array(obj, "tests")?,
    };
    let behavior = state.store.patch_behavior(&id, &patch, &ctx.actor)?;
    state.wake();
    Ok(Json(behavior.to_json()))
}

/// DELETE /v1/behaviors/{id} (write). Reported results stay.
pub async fn delete(
    State(state): State<Arc<AppState>>,
    Extension(ctx): Extension<AuthCtx>,
    Path(id): Path<String>,
) -> ApiResult<StatusCode> {
    ctx.require_scope("write")?;
    let existing = state.store.get_behavior(&id)?;
    ctx.require_project(&existing.project)?;
    state.store.delete_behavior(&id, &ctx.actor)?;
    state.wake();
    Ok(StatusCode::NO_CONTENT)
}

/// POST /v1/projects/{project}/runs (write) — report results. An optional
/// `Idempotency-Key` header makes a retried report record once.
pub async fn report(
    State(state): State<Arc<AppState>>,
    Extension(ctx): Extension<AuthCtx>,
    Path(project): Path<String>,
    headers: HeaderMap,
    ApiJson(body): ApiJson<Value>,
) -> ApiResult<impl IntoResponse> {
    ctx.require_scope("write")?;
    ctx.require_project(&project)?;
    let obj = body_object(&body)?;
    reject_unknown(obj, &RUN_FIELDS)?;
    let raw = match obj.get("results") {
        Some(Value::Array(items)) => items,
        _ => {
            return Err(ApiError::validation(
                "validation.run_results",
                "'results' must be an array of {test, outcome, detail?} objects.",
            )
            .remedy(
                "Send {\"results\": [{\"test\": \"<key>\", \"outcome\": \"pass\"}]}.".to_string(),
            ))
        }
    };
    let mut results = Vec::with_capacity(raw.len());
    for item in raw {
        let r = body_object(item)?;
        reject_unknown(r, &RESULT_FIELDS)?;
        results.push(ResultInput {
            test: require_str(r, "test")?,
            outcome: require_str(r, "outcome")?,
            detail: get_str(r, "detail")?,
        });
    }
    let req = RunReport {
        project,
        commit: get_str(obj, "commit")?,
        note: get_str(obj, "note")?,
        results,
        idempotency_key: idempotency_key(&headers)?,
        user: ctx.user.clone(),
    };
    let (mut out, replayed) = state.store.report_run(&req, &ctx.actor)?;
    state.wake();
    out["replayed"] = serde_json::json!(replayed);
    let status = if replayed {
        StatusCode::OK
    } else {
        StatusCode::CREATED
    };
    Ok((status, Json(out)))
}

/// GET /v1/projects/{project}/runs?limit=&offset= (read), newest first.
pub async fn list_runs(
    State(state): State<Arc<AppState>>,
    Extension(ctx): Extension<AuthCtx>,
    Path(project): Path<String>,
    RawQuery(raw): RawQuery,
) -> ApiResult<Json<Value>> {
    ctx.require_scope("read")?;
    ctx.require_project(&project)?;
    let pairs = query_pairs(raw.as_deref());
    let limit = parse_i64_param(&pairs, "limit")?;
    let offset = parse_i64_param(&pairs, "offset")?;
    let (items, total) = state.store.list_runs(&project, limit, offset)?;
    Ok(Json(super::paged(
        items,
        total,
        limit.unwrap_or(50).clamp(1, MAX_RUNS_PAGE),
        "Page with ?offset=N or raise ?limit= (max 200).",
    )))
}

/// GET /v1/projects/{project}/verification (read) — status counts overall and
/// per section, and reported tests no behavior links.
pub async fn summary(
    State(state): State<Arc<AppState>>,
    Extension(ctx): Extension<AuthCtx>,
    Path(project): Path<String>,
) -> ApiResult<Json<Value>> {
    ctx.require_scope("read")?;
    ctx.require_project(&project)?;
    let store = state.clone();
    let out = super::blocking_read(move || store.store.verification_summary(&project)).await?;
    Ok(Json(out))
}
